core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java

```diff
@@ -1634,71 +1674,221 @@ boolean resumeSyncForDocument(
+  }
 
-    final BsonDocument result = getLocalCollection(namespace)
-        .findOneAndUpdate(
-            getDocumentIdFilter(documentId),
-            update,
-            new FindOneAndUpdateOptions().returnDocument(ReturnDocument.AFTER));
+  /**
+   * Update a single document in the collection according to the specified arguments.
+   *
+   * @param filter        a document describing the query filter, which may not be null.
+   * @param update        a document describing the update, which may not be null. The update to
+   *                      apply must include only update operators.
+   * @param updateOptions the options to apply to the update operation
+   * @return the result of the update one operation
+   */
+  UpdateResult updateOne(
+      final MongoNamespace namespace,
+      final Bson filter,
+      final Bson update,
+      final UpdateOptions updateOptions) {
+    // read the local collection
+    final MongoCollection<BsonDocument> localCollection = getLocalCollection(namespace);
+
+    // fetch the document prior to updating
+    final BsonDocument documentBeforeUpdate = getLocalCollection(namespace).find(filter).first();
+
+    // if there was no document prior and this is not an upsert,
+    // do not acknowledge the update
+    if (!updateOptions.isUpsert() && documentBeforeUpdate == null) {
+      return UpdateResult.acknowledged(0, 0L, null);
+    }
 
-    if (result == null) {
+    // find and update the single document, returning the document post-update
+    final BsonDocument documentAfterUpdate = localCollection.findOneAndUpdate(
+        filter,
+        update,
+        new FindOneAndUpdateOptions()
+            .collation(updateOptions.getCollation())
+            .upsert(updateOptions.isUpsert())
+            .bypassDocumentValidation(updateOptions.getBypassDocumentValidation())
+            .arrayFilters(updateOptions.getArrayFilters())
+            .returnDocument(ReturnDocument.AFTER));
+
+    // if the document was deleted between our earlier check and now, it will not have
+    // been updated. do not acknowledge the update
+    if (documentAfterUpdate == null) {
       return UpdateResult.acknowledged(0, 0L, null);
     }
 
-    final ChangeEvent<BsonDocument> event =
-        changeEventForLocalUpdate(
-            namespace,
-            documentId,
-            ChangeEvent.UpdateDescription.diff(documentBeforeUpdate, result),
-            result,
-            true);
+    final ChangeEvent<BsonDocument> event;
+    final CoreDocumentSynchronizationConfig config;
+    final BsonValue documentId = BsonUtils.getDocumentId(documentAfterUpdate);
+
+    // if there was no document prior and this was an upsert,
+    // treat this as an insert.
+    // else this is an update
+    if (documentBeforeUpdate == null && updateOptions.isUpsert()) {
+      config = syncConfig.addSynchronizedDocument(namespace, documentId);
+      triggerListeningToNamespace(namespace);
```

Comment:
Curious, do you have a test that checks that if you were to remove this line, it would fail?

Reply:
I don't believe so– I can add a line in our new upsert tests that ensures that the stream is opened on the namespace.

Reply:
Added to `testUpsertOne` and `testUpsertMany` in `DataSynchronizerUnitTests` 👍